### PR TITLE
Select table number with NMEXT / read_nmext

### PR DESCRIPTION
**Pull request by kylebaron**
See #750 and #782

# Summary

Simple parsing of the extfile:

- Read the entire file.   
- Look for TABLE lines.   
- Return a list of values for `skip` and `nrows` that can be passed to `fread` or `read_table`.    
- Default value for index is `"last"` which means take the last table mapping.   
- Generate an error if a table index is requested that is greater than the number of tables found.   
- Special return if just one table is found (skip 1 and read infinity).   
- Some attributes are added to `read_nmext` output that identifies the table that was selected.   

-----------
# Todo

- [x] Write tests
- [x] Add example ext file with multiple targets

------------------

# Examples

```r
> code1 <- '
+ $NMEXT
+ run = 2005
+ project = system.file("nonmem", package = "mrgsolve"), 
+ index = 1
+ '
> 
> mod1 <- mcode("foo", code1)
Building foo ... done.
> 
> code1b <- '
+ $NMEXT
+ run = 1005
+ project = system.file("nonmem", package = "mrgsolve"), 
+ index = 111
+ '
> 
> mod1b <- mcode("foo", code1b)
 Error: [read_nmext] table 111 was requested, but only 1 tables were found in the ext file > 
> 
> codelast <- '
+ $NMEXT
+ run = 1005
+ project = system.file("nonmem", package = "mrgsolve"), 
+ index = "last"
+ '
> modlast <- mcode("foo2", codelast)
Building foo2 ... done.
> 
> code3 <- '
+ $NMEXT
+ run = 2005
+ project = system.file("nonmem", package = "mrgsolve"), 
+ index = 3
+ '
> 
> mod3 <- mcode("foo3", code3)
Building foo3 ... done.
> 
> param(mod1)

 Model parameters (N=11):
 name    value  . name   value
 THETA1  3.3    | THETA5 0.729
 THETA10 2.3    | THETA6 1.14 
 THETA11 0.0999 | THETA7 0.337
 THETA2  3.26   | THETA8 0.192
 THETA3  -0.611 | THETA9 0.692
 THETA4  -0.208 | .      .    
> param(modlst)
 Error in h(simpleError(msg, call)) : 
  error in evaluating the argument '.x' in selecting a method for function 'param': object 'modlst' not found > param(mod3)

 Model parameters (N=11):
 name    value  . name   value
 THETA1  3.31   | THETA5 0.752
 THETA10 2.31   | THETA6 1.14 
 THETA11 0.103  | THETA7 0.331
 THETA2  3.25   | THETA8 0.193
 THETA3  -0.614 | THETA9 0.699
 THETA4  -0.208 | .      .    
> 
> 
> read_nmext(path = "inst/nonmem/1005/1005.ext")
$param
$param$THETA1
[1] 9.50789

$param$THETA2
[1] 22.791

$param$THETA3
[1] 0.0714337

$param$THETA4
[1] 3.47451

$param$THETA5
[1] 113.277

$param$THETA6
[1] 1.02435

$param$THETA7
[1] 1.19212


$omega
           [,1]       [,2]       [,3]
[1,]  0.2138790  0.1207700 -0.0116278
[2,]  0.1207700  0.0945105 -0.0372064
[3,] -0.0116278 -0.0372064  0.0465631

$sigma
          [,1]     [,2]
[1,] 0.0491707 0.000000
[2,] 0.0000000 0.201769

$raw
$raw$ITERATION
[1] -1000000000

$raw$THETA1
[1] 9.50789

$raw$THETA2
[1] 22.791

$raw$THETA3
[1] 0.0714337

$raw$THETA4
[1] 3.47451

$raw$THETA5
[1] 113.277

$raw$THETA6
[1] 1.02435

$raw$THETA7
[1] 1.19212

$raw$SIGMA11
[1] 0.0491707

$raw$SIGMA21
[1] 0

$raw$SIGMA22
[1] 0.201769

$raw$OMEGA11
[1] 0.213879

$raw$OMEGA21
[1] 0.12077

$raw$OMEGA22
[1] 0.0945105

$raw$OMEGA31
[1] -0.0116278

$raw$OMEGA32
[1] -0.0372064

$raw$OMEGA33
[1] 0.0465631

$raw$OBJ
[1] 2405.916


attr(,"table")
[1] "TABLE NO.     1: First Order Conditional Estimation with Interaction: Goal Function=MINIMUM VALUE OF OBJECTIVE FUNCTION: Problem=1 Subproblem=0 Superproblem1=0 Iteration1=0 Superproblem2=0 Iteration2=0"
attr(,"index")
[1] 1
```

### CODE REVIEW

**Comment by kylebaron** [comment]
@timwaterhouse I'll also benchmark this with a bayes run; right now, it does this mapping regardless; if there is a slowdown with bayes, we can add an option to say there is only one table just read it in.

File: tests/testthat/test-nmxml.R [hunk 1]
```diff
@@ -279,6 +279,35 @@ test_that("NONMEM estimates from nmext", {
...
+  aa <- attributes(read_nmext(run = 2005, project = project, index = "last"))
+  expect_equal(aa$index,5)
+  expect_match(aa$table, "First Order Conditional")
+  
+  b <- mrgsolve:::nmext(run = 2005, project = project, index = 2)
+  ab <- attributes(read_nmext(run = 2005, project = project, index = 2))
+  expect_equal(ab$index,2)
+  expect_match(ab$table, "Importance Sampling")
+  
+  expect_true(a$theta$THETA3 != b$theta$THETA1)
```
Comment (timwaterhouse) [inline tests/testthat/test-nmxml.R:295]: Did you mean to compare different thetas from different tables here?

**Review by timwaterhouse — COMMENTED** [review]
Looks great, thanks Kyle. Just curious about that one test.

File: tests/testthat/test-nmxml.R (hunk 1 above)
Comment (kylebaron) [reply tests/testthat/test-nmxml.R:295]: Thanks; no, that's a mistake.

**Review by timwaterhouse — APPROVED** [review]

**Comment by kylebaron** [comment]
@timwaterhouse One more update; let the user pass `index = "single"` to just read all the rows.  This just lets the user revert to previous behavior for stuff like BAYES if needed.

**Review by timwaterhouse — APPROVED** [review]
Very helpful, thanks.

**Pull request merged**